### ISSUE TITLE
parse() raises when last line of input contains a statement

### ISSUE THREAD

**Issue by trotterdylan** [m-labs/pyparser#3]
I.e. if there's no trailing newline in the source code:

```
>>> pythonparser.parse('print "hola"')
<unknown>:1:13: fatal: unexpected eof: expected !=, %, &, (, *, **, +, -, ., /, //, <, <<, <=, <>, ==, >, >=, >>, @, [, ^, and, if, in, is, not, or, strbegin or |
print "hola"
            ^
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/trotterd/projects/grumpy/build/lib/python2.7/site-packages/pythonparser/__init__.py", line 58, in parse
    mode, flags, version, engine)
  File "/Users/trotterd/projects/grumpy/build/lib/python2.7/site-packages/pythonparser/__init__.py", line 28, in parse_buffer
    return parser.file_input(), lexer.comments
  File "/Users/trotterd/projects/grumpy/build/lib/python2.7/site-packages/pythonparser/parser.py", line 89, in outer_rule
    result = inner_rule(parser)
  File "/Users/trotterd/projects/grumpy/build/lib/python2.7/site-packages/pythonparser/parser.py", line 151, in rule
    parser.diagnostic_engine.process(error)
  File "/Users/trotterd/projects/grumpy/build/lib/python2.7/site-packages/pythonparser/diagnostic.py", line 165, in process
    raise Error(diagnostic)
pythonparser.diagnostic.Error: <unknown>:1:13: fatal: unexpected eof: expected !=, %, &, (, *, **, +, -, ., /, //, <, <<, <=, <>, ==, >, >=, >>, @, [, ^, and, if, in, is, not, or, strbegin or |
print "hola"
            ^
```

**Comment by whitequark** [comment]
This behavior is correct. The [Python grammar](https://docs.python.org/2/reference/grammar.html) has three entry points:

```
single_input: NEWLINE | simple_stmt | compound_stmt NEWLINE
file_input: (NEWLINE | stmt)* ENDMARKER
eval_input: testlist NEWLINE* ENDMARKER
[...]
stmt: simple_stmt | compound_stmt
simple_stmt: small_stmt (';' small_stmt)* [';'] NEWLINE
```

As you can see, a trailing newline is required in the input in every mode except `eval_input`. You can switch the mode using the `mode` argument to `pythonparser.parse` if desired.

**Comment by trotterdylan** [comment]
Huh weird. Can't argue with the grammar, but it does seem like pythonparser.parse() would be friendlier if it accepted input without trailing newlines. ast.parse() and exec both do:

```
>>> ast.parse('print "hola"')
<_ast.Module object at 0x10425df50>
>>> exec 'print "hola"'
hola
```

FWIW it looks like CPython hacks this in by replacing the ENDMARKER with a NEWLINE in the token stream: https://github.com/python/cpython/blob/master/Parser/parsetok.c#L213

Just something to think about.

**Comment by whitequark** [comment]
Disgusting. But I suppose I can do this.

**Comment by trotterdylan** [comment]
Yessss

**Comment by sbourdeauducq** [comment]
Uhm, WTF? Couldn't they change the grammar instead?

**Comment by sbourdeauducq** [comment]
@gvanrossum

**Comment by whitequark** [comment]
Fixed via #18.
